app/renderer/lib/client-frameworks/java.js

```diff
@@ -119,65 +105,72 @@ ${this.indent(code, 4)}
-                .addAction(
-                        FINGER.createPointerMove(
-                                Duration.ofMillis(0),
-                                PointerInput.Origin.viewport(),
-                                start.getX(),
-                                start.getY()))
-                .addAction(FINGER.createPointerDown(PointerInput.MouseButton.LEFT.asArg()))
-                .addAction(
-                        FINGER.createPointerMove(
-                                Duration.ofMillis(1000),
-                                PointerInput.Origin.viewport(),
-                                end.getX(),
-                                end.getY()))
-                .addAction(FINGER.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
-        driver.perform(Arrays.asList(swipe));
+final PointerInput FINGER = new PointerInput(PointerInput.Kind.TOUCH, "finger");
+Point start = new Point(${x1}, ${y1});
+Point end = new Point (${x2}, ${y2});
+Sequence swipe = new Sequence(FINGER, 1);
+swipe.addAction(FINGER.createPointerMove(Duration.ofMillis(0),
+    PointerInput.Origin.viewport(), start.getX(), start.getY()));
+swipe.addAction(FINGER.createPointerDown(PointerInput.MouseButton.LEFT.asArg()));
+swipe.addAction(FINGER.createPointerMove(Duration.ofMillis(1000),
+    PointerInput.Origin.viewport(), end.getX(), end.getY()));
+swipe.addAction(FINGER.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
+driver.perform(Arrays.asList(swipe));
   `;
   }
 
-  codeFor_getCurrentActivity () {
-    return `String activityName = driver.currentActivity()`;
+  // Execute Script
+
+  codeFor_executeScript (varNameIgnore, varIndexIgnore, scriptCmd, jsonArg) {
+    let assembledCommand;
+    if (jsonArg === undefined) {
+      assembledCommand = `driver.executeScript("${scriptCmd}");`;
+    } else {
+      // change the JSON object into a format accepted by Map.ofEntries:
+      // combine the keys and values into a flattened array, then stringify it
+      const argsValuesArray = [].concat(...(_.toPairs(JSON.parse(jsonArg))));
+      const argsValuesString = JSON.stringify(argsValuesArray).slice(1, -1);
+      assembledCommand = `driver.executeScript("${scriptCmd}", ImmutableMap.ofEntries(${argsValuesString}));`;
+    }
+    return assembledCommand;
   }
 
-  codeFor_getCurrentPackage () {
-    return `String packageName = driver.currentPackage()`;
+  // App Management
+
+  codeFor_startActivity (varNameIgnore, varIndexIgnore, ...args) {
+    const argNames = ['appPackage', 'appActivity', 'appWaitPackage', 'intentAction', 'intentCategory',
+      'intentFlags', 'optionalIntentArguments', 'dontStopAppOnReset'];
+    // zip argument names and values into tuples, but only if the value was provided
+    // then flatten the array of tuples and stringify it
+    const argsValuesArray = argNames.flatMap((k, i) => (args[i] === undefined) ? [] : [k, args[i]]);
+    const argsValuesString = JSON.stringify(argsValuesArray).slice(1, -1);
```

Comment:
maybe we could dedupe this into a separate function and at the same time make it more readable

Reply:
I'll just rewrite this so that it reuses `codeFor_executeScript`